torch/utils/data/datapipes/iter/combinatorics.py

```diff
@@ -86,24 +85,24 @@ def __init__(self,
         else:
             self.datapipe = datapipe.unbatch(unbatch_level=unbatch_level)
         self.buffer_size = buffer_size
-        self._buffer = []
 
-    def buffer_replace(self, x):
-        idx = random.randint(0, self.buffer_size - 1)
-        val = self._buffer[idx]
-        self._buffer[idx] = x
+    @staticmethod
+    def buffer_replace(buffer, x):
+        idx = random.randint(0, len(buffer) - 1)
+        val = buffer[idx]
+        buffer[idx] = x
         return val
 
     def __iter__(self) -> Iterator[T_co]:
-        # TODO: Buffer is global, should be per __iter__ !!!
+        buffer: List[T_co] = []
```

Comment:
Also a nit for a use case that is probably uncommon and not be worth changing.
If the inputs tend to repeat, not-randomly ordered, and the buffer size is small relative to input size, you may want to have the buffer be a `dict` of `key: count`, then use `random.sample` to draw one at a time.

I think real world ML data sets are unlikely to have repeated and exact inputs though, so this will not matter.

Reply:
I think this is too micro-optimization. I will leave it for users who want to implement their own buffer shuffle based on their input data.

Reply:
Verified that TF is using the same strategy that non-shuffling when buffer is 1. So, I will leave it as it is.